File: python/lite_sdk2/topics.py
from .messages.low_command import LowCommand
from .messages.low_state import LowState


LOWCOMMAND = "/lowcommand"
LOWSTATE = "/lowstate"

_DEFAULTS: dict[type, str] = {
    LowCommand: LOWCOMMAND,
    LowState: LOWSTATE,
}


def default_topic(message_type: type) -> str:
    try:
        return _DEFAULTS[message_type]
    except KeyError as exc:
        raise ValueError(f"No default topic registered for {message_type.__name__}.") from exc


def ros_topic_to_dds(name: str) -> str:
    if name.startswith("rt/"):
        return name
    return f"rt{_normalize_ros(name)}"


def dds_topic_to_ros(name: str) -> str:
    if not name:
        raise ValueError("DDS topic names must be non-empty.")
    if name.startswith("rt/"):
        return name[2:]
    return name


def _normalize_ros(name: str) -> str:
    if not name:
        raise ValueError("ROS topic names must be non-empty.")
    if name.startswith("rt/"):
        raise ValueError("Expected a ROS topic name, not a DDS topic name.")
    return name if name.startswith("/") else f"/{name}"

File: rust/src/topics.rs
//! Topic name constants and ROS↔DDS conversion helpers.
//!
//! Mirrors the Python `lite_sdk2.topics` module so both sides agree on
//! topic naming (ROS names with a `/` prefix get a `rt/` prefix on the wire).

pub const LOWCOMMAND: &str = "/lowcommand";
pub const LOWSTATE: &str = "/lowstate";

pub fn ros_topic_to_dds(name: &str) -> String {
    if name.starts_with("rt/") {
        return name.to_string();
    }
    if name.starts_with('/') {
        return format!("rt{name}");
    }
    format!("rt/{name}")
}

pub fn dds_topic_to_ros(name: &str) -> &str {
    // Matches Python: `rt/lowstate` → `/lowstate` (keeps the leading slash).
    name.strip_prefix("rt").unwrap_or(name)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn ros_absolute_to_dds() {
        assert_eq!(ros_topic_to_dds("/lowcommand"), "rt/lowcommand");
    }

    #[test]
    fn ros_relative_to_dds() {
        assert_eq!(ros_topic_to_dds("lowcommand"), "rt/lowcommand");
    }

    #[test]
    fn already_dds_passes_through() {
        assert_eq!(ros_topic_to_dds("rt/lowcommand"), "rt/lowcommand");
    }

    #[test]
    fn dds_to_ros_strips_prefix() {
        assert_eq!(dds_topic_to_ros("rt/lowstate"), "/lowstate");
    }
}
